hal/src/common/thumbv6m/sercom/v2/spi.rs

```diff
@@ -1428,6 +1444,22 @@ impl<C: ValidConfig> Spi<C> {
         &self.config.as_ref().sercom()
     }
 
+    /// Update the SPI bus baudrate.
+    ///
+    /// **WARNING** this method assumes that the GCLK clock hasn't been changed
+    /// since [`Spi`] has been initialized.
+    ///
+    /// Calling this method will temporarily disable the SERCOM peripheral, as
+    /// the BAUD register is enable-protected. This may interrupt any ongoing
+    /// transactions.
```

Comment:
Maybe it would be a good idea to wait for `TXC` to not interrupt ongoing transactions?`
I don't think updating the baud rate will ever be that urgent that you can't at least wait for one byte/word to be sent to prevent data corruption.

Reply:
That's reasonable

Reply:
This is a good question. I think my answer in this case is no, but while I have your attention, I'd like to ask your opinions on a related question.

In general, the `Spi` struct does not track the transaction status. Without tracking that status, I don't see how you could implement this suggestion. And personally, I don't think the use case is common enough to justify adding the overhead of tracking transaction status. I think it's something users should do for themselves, when necessary.

Note, however, that there is one case where I'm already tracking transaction status, the implementation of `serial::Read` for read-only SPI masters (i.e. MISO but no MOSI). There, I track the transaction status so I know when to start a transaction vs. wait for the response. It would be conceivable to expand that implementation and always track the state, but I don't particularly like it. Thoughts?

Separately, I would like to ask you about a philosophical question regarding the `TXC` flag and the `DATA` register. When I wrote this module, I purposefully made the `read_data` and `write_data` methods `unsafe`, because directly reading the `DATA` register could violate some assumptions of the `embedded-hal` trait implementations. However, I wasn't very principled, because I allowed the `clear_flags` method to be marked safe, even though clearing the `TXC` flag could potentially disrupt the implementation of `serial::Write::flush`.

I haven't really come to a coherent conclusion on this point. Strictly speaking, I think there is an argument to be made that the `read_data` and `write_data` methods are not memory unsafe. Calling them incorrectly could lead to a logic error, but I don't see any immediate way to produce memory unsafety. Do you agree? Should I mark these functions safe? Or, alternatively, do you think `clear_flags` should be marked `unsafe`, because it could disrupt an `embedded-hal` trait? Or is the status quo sufficient?

Reply:
I agree that waiting for a TXC might be a little overkill considering the use cases of this API. I don't really see the point of drastically changing the implementation just to support this functionality.

On the topic of marking `read_data` and `write_data` as `unsafe`, I would say it's definitely a grey area. It's not exactly memory unsafe, and `unsafe` shouldn't be abused to cover logic errors, but it can be seen as a bit of a pseudo-thread-safety problem.

Manually clearing the flags (or manually reading/writing the DATA register, which manipulates the flags), is essentially mutating a `static mut` variable, which can be mutated asynchronously by different contexts (main context, interrupt handlers, or the hardware itself). For that reason, I think it's sound to leave these functions as `unsafe`.